### ISSUE TITLE
RunTimeWarning is sometimes raised in vectorized function and should not

### ISSUE THREAD

**Issue by cderemble** [numba/numba#3559]
The following vectorized function raises a warning saying it encountered an invalid value despite the fact there is a check in the function to make sure to skip `nan` values. The expected behaviour is not warning raised.

```
import numpy
import numba
import warnings
warnings.simplefilter("always", RuntimeWarning)


def test_func(value):
    if numpy.isnan(value):
        return numpy.nan
    if value <= 0.0:
        return 1.0
    return 0.0


numba_func = numba.vectorize([numba.float64(numba.float64)])(test_func)
numpy_func = numpy.vectorize(test_func)


a = numpy.full(1, numpy.nan)
b = numpy.full(32, numpy.nan)

print(numba_func(a))  # no warning: ok
print(numba_func(b))  # warning: invalid value encountered
print(numpy_func(a))  # no warning: ok
print(numpy_func(b))  # no warning: ok
```
The output of this script is
```
[nan]
test_numba.py:23: RuntimeWarning: invalid value encountered in test_func
  print(numba_func(b))
[nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
 nan nan nan nan nan nan nan nan nan nan nan nan nan nan]
[nan]
[nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
 nan nan nan nan nan nan nan nan nan nan nan nan nan nan]
```

**Comment by stuartarchibald** [comment]
Thanks for the report. I don't seem to be able to reproduce this locally. Please could you share the following to help debug it (the `numba -s` command shows all this):

* Numba version
* NumPy version
* Python version
* Platform (OS and architecture)

Thanks.

**Comment by cderemble** [comment]
I am seeing it both on Linux (x86_64) and Windows 10 (x64-based), using numba 0.40 and 0.41
I am using a conda environment that contains this list:
```
blas                      1.0                         mkl
ca-certificates           2018.03.07                    0
certifi                   2018.10.15               py37_0
intel-openmp              2019.1                      144
libedit                   3.1.20170329         h6b74fdf_2
libffi                    3.2.1                hd88cf55_4
libgcc-ng                 8.2.0                hdf63c60_1
libgfortran-ng            7.3.0                hdf63c60_0
libstdcxx-ng              8.2.0                hdf63c60_1
llvmlite                  0.26.0           py37hd408876_0
mkl                       2018.0.3                      1
mkl_fft                   1.0.6            py37h7dd41cf_0
mkl_random                1.0.1            py37h4414c95_1
ncurses                   6.1                  he6710b0_1
numba                     0.41.0           py37h962f231_0
numpy                     1.15.4           py37h1d66e8a_0
numpy-base                1.15.4           py37h81de0dd_0
openssl                   1.1.1a               h7b6447c_0
pip                       18.1                     py37_0
python                    3.7.1                h0371630_3
readline                  7.0                  h7b6447c_5
setuptools                40.6.2                   py37_0
sqlite                    3.25.3               h7b6447c_0
tk                        8.6.8                hbc83047_0
wheel                     0.32.3                   py37_0
xz                        5.2.4                h14c3975_4
zlib                      1.2.11               h7b6447c_3
```

**Comment by stuartarchibald** [comment]
Great, thanks. I still cannot reproduce on linux but can on OSX, I'll take a look.

**Comment by seibert** [comment]
It is puzzling, since it seems to depend on the behavior of the DUfunc, since only the version without an explicit type signature seems to trigger the issue, and only on the first compilation pass?  Also strangely, the warning goes away if the ufunc body is changed to:
```
def test_func(value):
    return numpy.nan
```
i.e., the warning is not triggered by returning a NaN in the output array.  My initial theory was that this had to do with our implementation of `numpy.isnan` leaving a floating point exception flag set for NumPy to find, but I don't see how that lines up with the behavior of seeing the warning only on the first call, when the ufunc is compiled.

**Comment by cderemble** [comment]
actually, I see the warning on every call but only if the length of the input is big enough (32 on my linux box, but only 16 on my windows box).

**Comment by bsamuel-ui** [comment]
I was running something similar and got spurious warnings:

```
@vectorize
def dbz(num, den):
    if den != 0.0:
        return num / den
    else:
        return 0.0
```

But.... if you insert `print(num, den)` right before `return num / den`, no warning is issued. If you set `NUMBA_OPT=0`, again no warning is issued. I can put up my configuration if it would add anything.

Another point: `with errstate(invalid='ignore'):` about doubles the performance, so those warnings are pretty expensive.

**Comment by gokceneraslan** [comment]
I also have this issue. Here is my code to reproduce:

```python
import numpy as np
from numba import vectorize, float64, float32, int32, int64

@vectorize([float64(int32),
            float64(int64),
            float32(float32),
            float64(float64)])
def log1p(x):
    # see: https://github.com/numpy/numpy/blob/master/numpy/core/src/npymath/npy_math_internal.h.src#L86
    if np.isinf(x) and x > 0: 
        return x
    
    u = 1. + x
    d = u - 1.
    
    if d == 0.:
        return x

    return np.log(u) * x / d

x = np.array([[0.      , 0.      , 0.      , 0.      ],
              [0.      , 3.912363, 3.912363, 0.      ]], dtype=np.float32)

x, log1p(x)
```

**Comment by esc** [comment]
FWIW this seems to be related: https://github.com/numba/numba/issues/4793

**Comment by stuartarchibald** [comment]
Added explanation here: https://github.com/numba/numba/issues/4793#issuecomment-622323686

**Comment by esc** [comment]
@stuartarchibald can we mark this as a duplicate of #4793 and close?

**Comment by stuartarchibald** [comment]
Probably worth checking the machine code to make sure it is actually the same problem, then close.

**Comment by stuartarchibald** [comment]
> @esc please could you see if this still happens and dump the machine code here if it does? Thanks!

I've looked at the machine code, seems like it's a variant on https://github.com/numba/numba/issues/4793#issuecomment-622323686. `@vectorize` naturally generates various sized "vector" loops to ensure wider instructions are targetted when more data is present. I think the variation on whether a signal is raised depends on whether a scalar comparison or vector comparison instruction is in use. Closing.
